### DISCUSSION TITLE
Identity and authorization management in Xomega solution

### DISCUSSION THREAD

**Discussion by osd1syc** [Xomega-Net/Xomega.Net4VS#22] [General]
How can we integrate asp net core identity to the xomega solution so that it can work as central identity and authorization management solution? Or any other open source library for authentication and authorization management ?

**Comment by XomegaNet** [comment]
Thanks for your questions and your interest in Xomega.

If you [watch our latest videos](https://youtu.be/X6DWEK6BkuU), then you'll see that the Xomega solutions that you create with our solution wizard are preconfigured to support password-based authentication with a login form, as well as claims-based authorization control. It's designed to support whatever architecture you select in the wizard - from cookie-based authentication on the server side for Blazor Server solutions to JWT-based API security for Blazor WebAssembly, MAUI or WPF solutions. Do you have a specific architecture in mind that you want to use?

The default Xomega solution provides all the plumbing for the application security, but it's not plugged into any identity and authorization management system. You have an option to either implement your own solution or use one of the standard implementations, such as the ASP.NET Core Identity. For example, our tutorial shows you [how to implement security](https://xomega.net/docs/tutorial/security/overview) using existing tables from the sample database AdventureWorks from Microsoft. Basically, all you have to do is the following:
- [Implement password validation](https://xomega.net/docs/tutorial/security/authentication#password-validation) in the login operation.
- [Implement reading of the current UserInfo](https://xomega.net/docs/tutorial/security/authentication#reading-user-info) with all the security data.
- [Implement converters between the UserInfo and the ClaimsPrincipal](https://xomega.net/docs/tutorial/security/authentication#user-info-principal-converter) and vice versa.

If you want to use the ASP.NET Identity instead of custom tables, then you can just implement the above using the standard identity entities and services. Note that ASP.NET Identity can provide you some UI for editing the user profile and managing users, but it will use plain ASP.NET Core razor pages that may look different from the rest of your Xomega application. Does that work for you?

Please let me know if you have any further questions or need help with this.

**Reply by osd1syc** [reply]
Hi, 

Thanks for very prompt reply. Since basic user related plumbing is currently set and is everywhere in all projects, integrating with aspnet core identity solution or any other authentication and authorization library is going to be painful to take care of all hooks every where. Is there an easy way to remove currently set authentication and to add an Authentication project as a separate application tier (Which can not only authenticate the user but can also authorize API access points. Here application approach is to have blazor, desktop, maui all application use API to access data layer and API project depends on this to be added authentication and auth project. Log in, register interface can be managed using redirect scripts.) ? Can you list out steps to achieve this? This will provide additional benefit that developers will be free to use any authentication and authorization solution since it will sit separate from all other projects. Thanks

**Reply by XomegaNet** [reply]
Thanks for the additional info. Before listing the steps, I would like to provide more details on Xomega, as well as some options for you, in order to make sure I fully understand the exact security configuration that you're looking for.

> Which can not only authenticate the user but can also authorize API access points.

FYI, Xomega security provides not only authentication, but also authorization checks, which we encourage to perform in the business services, as described [in our tutorial](https://xomega.net/docs/tutorial/security/secure-service#restricting-access-to-operations). This will make it work regardless of whether you expose your services via a REST API, gRPC, WCF or call directly from the presentation logic, such as in the Blazor Server configuration. That being said, you're still free to add authorization checks to the API access points in the generated controllers, if you want to.

> Here application approach is to have blazor, desktop, maui all application use API to access data layer

Given that you want to include desktop and MAUI applications, I assume that your API would have to be secured using JWT rather than cookies, since using cookies in non-browser applications can be problematic. Let me know if this is not the case.

> add an Authentication project as a separate application tier

If you build the authentication and authorization management into the REST API of your Xomega solution, then it will be limited to your application only. Is your goal to have it as a separate server (an identity provider) so that other applications and APIs would also be able to use it for authentication and authorization?

If so, then one option could be to just create a separate ASP.NET Core project for it, move the `AuthController` from the `.Services.Rest` project to that new project, and implement the `IPasswordLoginService` in that project using ASP.NET Core Identity or any other identity framework of your choice. This will allow your UI to still use all the security plumbing to be able to login and communicate to the REST API. You'll just need to provide a different URL for the login/refreshToken operations to point to the new identity server. Is it something that would work for you?

> Since basic user related plumbing is currently set and is everywhere in all projects, integrating with aspnet core identity solution or any other authentication and authorization library is going to be painful to take care of all hooks every where. Is there an easy way to remove currently set authentication

If you don't want any of the Xomega default authentication infrastructure, then it should not be that hard to remove it. It will mostly consist of removing it from the main `Program.cs` files and deleting certain security-related files that are no longer needed. I can provide steps for doing this, but we can also include an option in the solution wizard to generate a solution with no auth support. Would you find such an option useful?

Regardless, with that security code removed, the question will be on what to replace it with. If you are looking to use a third-party identity provider, then you will need to add the code, as recommended by that provider. If you go with something that supports standard OAuth2/OIDC, then you can use the MSAL library for authentication with such a provider. As far as integrating it with Xomega at this point, you'll just need to configure an implementation of the `IPrincipalProvider`, such as `AuthStatePrincipalProvider` on the Blazor client side and `ContextPrincipalProvider` on the API side.

With all this being said, could you please let me know which of these options works best for you, or if you're looking for something else, so that I could provide you with further info?

**Reply by osd1syc** [reply]
Thanks for the prompt and so detailed reply. I appreciate that you have tried to respond to and conveyed your willingness to help for all possible Authentication and Authorization options. Of various options, since value proposition of Xomega framework is to generate SPA, desktop, Web (Blazor) and now Maui app (Cross platform development framework) with strong code generation capabilities, the identity and authorization management needs to be centrally managed else individual applications at different stages of development will have versioning challenges in managing identity and access control. I would prefer to have rest API project manage all transactions and business logic with different projects taking care of only respective UIs. 

That would give us option of having identity tightly wrapped with Rest API project or having a separate identity project. Between the two options, I would recommend to have identity as a separate project because then one can not only use Microsoft identity platform, but can use other open source and commercial identity providers. This would also help larger enterprises manage identity with SSO feature. At the stage of code generation, in the generator templates, one can set Authorize attribute either true or false where upon the generator can control how the controllers are annotated. Having identity as a separate project would allow developers to decide on RBAC or CBAC or PBAC as per project need. 

It would help if you could guide about how to remove all user related plumbing from current version and how to add aspnet core identity as a separate API project with provision for login, register, log out, JWT tokens and identity pages in respective UIs (That can be part of xomega framework project template)  that connect with identity APIs. 

I sincerely appreciate your work. Thanks

**Reply by XomegaNet** [reply]
I went through the projects for the main modern Xomega apps (Blazor, MAUI, WPF), and created a list of steps to remove all authentication and authorization code from them. As I said before, we can also add an option to the solution wizard to create the solution with no auth (would you find it useful?), but here are the manual steps.
1. Delete the Auth folders from all the projects as follows.
`.Model` > Auth
`.Client.Blazor.Common` > Views/Auth
`.Client.Common` > DataObjects/Auth and ViewModels/Auth
`.Client.Wpf` > Views/Auth
`.Services.Common` > ServiceContracts/Auth
`.Services.Entities` > Services/Auth
1. Build the `.Model` project to regenerate all artifacts
1. Perform the following updates in the `.Client.Blazor.Common` project.
Delete `UrlConfig.cs`, `AuthConfig.cs` and `Components/LoginDisplay.razor` files.
Remove `[Authorize]` attribute from the `Views/Home.razor`.
Update `Components/MainLayout.razor` to remove the `LoginDisplay` component.
Update `Routes.razor` to replace `AuthorizedRouteView` with `RouteView` and remove the nested `NotAuthorized` element.
Update `_Imports.razor` to remove a using for `.Client.Blazor.Common.Views.Auth` namespace
1. Perform the following updates in the `.Client.Blazor` project.
Delete `ServerLoginNavigation.cs` and `SignInEndpoints.cs` files.
Update `Program.cs` to remove code from setting of the `apiAwareRedirect` up until the `AddRazorComponents` call, remove adding of `AuthorizeFilter`
Also remove calls to `AddSignIn` and `AddSignOut` as well as a using for `UrlConfig` in that file.
1. Perform the following updates in the `.Client.Wasm` project.
Delete `WasmLoginNavigation.cs` file.
Update `Program.cs` to remove code under the Auth configuration comment and unused usings. Add `services.AddAuthorizationCore()`.
Update `appsettings.json` to set "Authorization" to "false" on the "RemoteApi", as applicable.
1. Perform the following updates in the `.Client.Maui` project.
Delete `MauiLoginNavigation.cs` and `Components/SignOut.razor` files.
Update `MauiProgram.cs` to remove code under the Auth configuration comment (except for `AddAuthorizationCore`) and unused usings
Update `appsettings.json` to set "Authorization" to "false" on the "RemoteApi".
1. Perform the following updates in the `.Client.Wpf` project.
Delete `LoginNavigation.cs` and `TokenService.cs` files.
Update `WpfAppInit.cs` to remove registration of `IPrincipalConverter`, `ILoginNavigation`, `IPasswordLoginService` and `ITokenService`
Update `App.xaml.cs` to replace `await LoginViewCustomized.PopupLoginAsync()` with `MainView.Start()`.
Update `MainViewWindowCreator.cs` to remove a check for `LoginView`.
Update `appsettings.json `to set "Authorization" to "false" on the "RemoteApi".
1. Perform the following updates in the `.Services.Rest` project.
Delete `AuthController.cs` file.
Update `Program.cs` to remove code under the Auth configuration comment and unused usings, remove adding of `AuthorizeFilter`

Once you clean up the auth code from the Xomega solution, you can easily add a new ASP.NET Core Web App with Individual Accounts authentication to your solution, which will add ASP.NET Core Identity. However, I'm not quite sure how exactly you wanted to integrate it into the authentication mechanism for all other projects. There are several things to keep in mind regarding ASP.NET Core Identity:
- Microsoft recommends using cookie authentication with ASP.NET Core Identity rather than tokens. However, it's not clear how that would work for non-browser clients, such as MAUI and WPF.
- By default, ASP.NET Core Identity cannot redirect to a non-local URL after the login to [prevent redirect attacks](https://learn.microsoft.com/en-us/aspnet/core/security/preventing-open-redirects?view=aspnetcore-9.0). This will be a problem for browser apps that use a separately hosted ASP.NET Core Identity project.
- ASP.NET Core Identity does not support JWT. For token authentication it uses proprietary Bearer tokens, which would not be compatible with other identity providers.

Do you have any reference sample solutions that use a separately hosted ASP.NET Core Identity for authentication of Blazor (Server and WebAssembly), MAUI and WPF with a REST API?

**Reply by osd1syc** [reply]
Hi, sorry for being late. I followed all steps, removed entire Authentication and authorization plumbing. However, i realize that having identity as a separate project as part of solution would not work because for all projects (MAUI, SPA) we will have to create models/view models and client service to handle identity. That makes project very complicated and almost impossible to manage. Stripping all auth code, while i try to run the application, it still breaks due to passing of Cancellation Token in each controller/service methods. 

As an alternative, have you given thought to having say Keycloak manage identity and authorisation ?If the template can provide that option while generating code to use Keycloak or any other Oauth platform and create necessary hooks or if you can provide a guide/walkthrough on integration with Keycloak then that would be great.  Thanks

**Reply by XomegaNet** [reply]
Thanks for the update. In the next release of Xomega.Net we are going to add a selection of authentication options to the solution wizard. In addition to the current password-based authentication, where you supply custom code for password validation and claims creation, it will have an option to use ASP.NET Core Identity, which will use existing ASP.NET Identity tables for such validation. Those tables can be part of your application or in a separate database, if you want to share it between multiple applications. You can then add a standard standalone web application with Identity to allow the users to manage their profiles via a standard web interface.

Additionally, there will be an option for no authentication, which will create a solution with no login screens or any of the auth plumbing. This will allow you to integrate other 3rd party identity providers manually, as needed. You will mostly need to follow the steps recommended by the specific provider for your type of app (Blazor Server, WebAssembly, MAUI, WPF, etc.).

Later on, we may add other authentication options, such as Microsoft Entra or other OAuth/OIDC providers.

**Reply by osd1syc** [reply]
That would be good. I would like to give further following feedback if it can be taken care of:
1. After removing auth code, I could not run application. It seems that for some strange reason, Blazor, Rest API etc projects were not able to pick up connection string from db.config in Entity project. I had to manually add connection string to appsettings.json. It would be better to add connection string to respective projects during code generation. Remotely Blazor/Maui/SPA can interact through Rest API project which would have project dependency with Entity project but Blazor/SPA/Maui may not have that dependency if they run through Rest API. 
2. In Rest project, Swagger is not added. To access end points, I had to add Swagger and Swagger UI. If it is preconfigured then it would be easier to test APIs. 
3. Rest API project does not have pagination in controllers (Though it has lazyloading proxies enabled). I could not get list view of a table with 10000 records, with nested records as sub object, display in browser. It would be desirable to have pagination enabled for APIs also. (Blazor UI does it at UI level). Generator config has options for Blazor UI but not for Rest Project or SPA  project.
4. Currently Blazor, SPA run with entities, services etc independently contained in their own projects. If they interface with data layer through Rest API project then it would be easier to add health check, logging, audit etc with Rest project and keep clients lean.
5. While running and testing my blazor as well as Rest APIs, i found application to be very efficient with data operations. My VS showed memory utilization go up by 10-20 mb for every list view load. However, after the data was loaded in views, connection/thread was not releasing memory and i could see memory utilisation go up in step function manner. I would suggest you to examine memory leak/ threads not closing connections after every transaction. 
6. While running list views, it first offers search screen and only when you do search with parameter or no parameter that the records load. Default behaviour should be to load list view without any search applied. 
7. Regarding Auth project mentioned by you, I would suggest to make it a stand alone project with preconfigured code in blazor, SPA, MAUI for exchange of JWT /cookies through rest API to make one backend work for all frontends. This would need to be independently hosted.  Including project while building and publishing would work for blazor but would not work for SPA/MAUI. 

All in all, thanks.

**Reply by XomegaNet** [reply]
Thanks so much for your feedback. I'll try to answer it for each item, but it may make more sense to create a new discussion for each individual item to further continue the conversation over there.

> 1. After removing auth code, I could not run application. It seems that for some strange reason, Blazor, Rest API etc projects were not able to pick up connection string from db.config in Entity project...

The startup projects that require a DB connection, such as Rest and Blazor (for Blazor Server), have a post-build event to copy the `db.config` over from the Entities project, so that it would be accessible for each such project. **Did that not work for you?** We can revisit how the users configure their DB connection, but it seems pretty easy for them to update it themselves. Normally DB connection string should be stored in the user secrets for development, so that it's not checked into the source control, and it should be provided by the environment in the deployed app.

> 2. In Rest project, Swagger is not added. To access end points, I had to add Swagger and Swagger UI. If it is preconfigured then it would be easier to test APIs.

You can also test API using the new .http files in VS. It should not be that hard for the users to add their preferred Swagger implementation, but we can definitely include one by default, if you find it useful.

> 3. Rest API project does not have pagination in controllers (Though it has lazyloading proxies enabled). I could not get list view of a table with 10000 records, with nested records as sub object, display in browser. It would be desirable to have pagination enabled for APIs also...

Currently pagination is performed client-side, and is not included in the service operations by default (unless you add it as a new parameter, and implement it manually). If you have a large table, you can add custom code to limit the maximum number of records returned to a certain number, e.g. 3000, so as to not overwhelm the service. (Let me know if you need help with this.) Normally, the users should specify some criteria that would make it return fewer records than that maximum though. **Do you have a use case where you need to return all 10000 records to the client?**

That being said, we are looking into ways to improve "read list" operations, which would include support for pagination as well.

> 4. Currently Blazor, SPA run with entities, services etc independently contained in their own projects. If they interface with data layer through Rest API project then it would be easier to add health check, logging, audit etc with Rest project and keep clients lean.

I'm not sure I fully understand this, but client projects that use Rest API should not have a dependency on the Entities project. They only have dependency on the `Services.Common` project, which contains DTOs and service contracts, and is shared between the client and the services. If your client needs to access the DB directly rather than through a Rest API, such as for Blazor Server or 2-Tier MAUI/WPF apps, then those projects would have a dependency on the Entities project.

> 5. While running and testing my blazor as well as Rest APIs, i found application to be very efficient with data operations. My VS showed memory utilization go up by 10-20 mb for every list view load...

We will try to check for memory leaks, but if you have any specific steps that you can share that cause memory leaks, it would be great if you could open a separate issue and describe them in there.

> 6. While running list views, it first offers search screen and only when you do search with parameter or no parameter that the records load. Default behaviour should be to load list view without any search applied.

Always running a blank search on a list view may cause performance issues on large datasets, therefore it's not enabled by default. It makes sense only if the default results will be highly relevant for the users. In that case, you can configure the URL for the menu option that opens the search screen, and add query parameters to set up values for the initial criteria. After opening the screen, the users will be able to review those default criteria and either modify them or hit Search. If you think the default criteria will be used most of the time, then you can also add the `_action=search` query parameter to auto-run the Search upon opening the view. You can [check this doc](https://www.xomega.net/docs/framework/blazor/views#page-component) for an example.

> 7. Regarding Auth project mentioned by you, I would suggest to make it a stand alone project with preconfigured code in blazor, SPA, MAUI for exchange of JWT /cookies through rest API to make one backend work for all frontends...

The app projects in the new release will use ASP.NET Core Identity framework and DB to perform authentication and issue JWT/cookies, but you will be able to add a separate standard ASP.NET Core web app with Individual Accounts and point it to the same identity DB to allow users to manage their passwords and profiles via a standalone web interface.

**Comment by XomegaNet** [comment]
FYI, the latest release 9.12.5 of Xomega.Net now allows you to select an authentication option when creating a Xomega solution. Available authentication options are:

- **Password (Custom)** - this is the custom password-based security that was included in the previous version by default.
- **ASP.NET Core Identity** - the password-based authentication using standard ASP.NET Core Identity framework and entities/tables.
- **None** - no authentication support by default, which allows you to manually integrate with any identity provider of your choice.

Please check [our docs](https://xomega.net/docs/visual-studio/new-projects/new-solutions#selecting-solution-components) for any further details and let us know if you have any questions.
